l2-contracts/contracts/ConsensusAuthority.sol

```diff
@@ -0,0 +1,92 @@
+// SPDX-License-Identifier: MIT
+
+pragma solidity 0.8.20;
+
+import "./ValidatorRegistry.sol";
+import "./AttesterRegistry.sol";
+
+contract ConsensusAuthority {
+    // Owner of the contract (i.e., the governance contract).
+    address public owner;
+    // The validator registry instance.
+    ValidatorRegistry public validatorRegistry;
+    // The attester registry instance.
+    AttesterRegistry public attesterRegistry;
+
+    // A map of owningPubKey (a cold key L2 address) => indexes on registries.
+    mapping(address => RegistryIndexes) public nodes;
+
+    struct RegistryIndexes {
+        // Index in the validator registry.
+        uint256 validatorIdx;
+        // Index in the attester registry.
+        uint256 attesterIdx;
+    }
+
+    modifier onlyOwner() {
+        require(msg.sender == owner, "Unauthorized: onlyOwner");
+        _;
+    }
+
+    modifier onlyOwnerOrNodeOwner(address owningPubKey) {
+        require(msg.sender == owner || msg.sender == owningPubKey, "Unauthorized: onlyOwnerOrNodeOwner");
+        _;
+    }
+
+    constructor(address _owner) {
+        owner = _owner;
+        validatorRegistry = new ValidatorRegistry(address(this));
+        attesterRegistry = new AttesterRegistry(address(this));
+    }
+
+    // Adds a new node to the registries.
+    function add(
+        address owningPubKey,
+        uint256 weight,
+        bytes calldata validatorPubKey,
+        bytes calldata validatorPoP,
+        bytes calldata attesterPubKey
+    ) external onlyOwner {
+        uint256 validatorIdx = validatorRegistry.add(weight, validatorPubKey, validatorPoP);
+        uint256 attesterIdx = attesterRegistry.add(weight, attesterPubKey);
+        nodes[owningPubKey] = RegistryIndexes(validatorIdx, attesterIdx);
+    }
+
+    // Inactivates a node.
+    function inactivate(address owningPubKey) external onlyOwnerOrNodeOwner(owningPubKey) {
+        RegistryIndexes memory node = nodes[owningPubKey];
+        validatorRegistry.inactivate(node.validatorIdx);
+        attesterRegistry.inactivate(node.attesterIdx);
```

Comment:
IIUC the registry `remove` methods change the index of a validator/attestor in the array, by moving the last one to the empty space. How does htis get reflected in the `RegistryIndexes`?

Reply:
Ah, I guess like `inactivate` affects both, a future implementation of `remove` will remove both as well, right?

Reply:
The unique ID referred to as the index (or just `idx`) is different than the element index in indexes array. I've referred to the latter `offset`, because as you mentioned, it is ephemeral (due to gas-optimized element removal). But the unique ID, stored at the top-level contract, is fixed. 

This might still be confusing, so perhaps it's better to refer to the former simply as UID. 

@brunoffranca any thoughts?

Reply:
Ah, you're right, thanks 🙈 
I mixed up the `attesterIndexes` and `attesters[idx]` thing and started thinking that `attesters` is an array 👍 

My first thought when I saw this design, instead of a `mapping(address => ...)` was that it might be more efficient to store in an array, and that made me forget that it's actually a `mapping(uint256 => ...)`. I am guessing the reason for this is that you can change the public key of a node in this contract, which means the real key is the node's `owningPubKey`. Could _that_ be the key inside the `AttesterRegistry` and the `ValidatorRegistry` instead of the sequence number?

Reply:
No opinion here. If you think UID is less confusing, go ahead.

Reply:
@aakoshh I agree. The only implication I can see is that the node owner address will be stored on the registries, which I don't see a problem with, since it can already be traced back from the validator/attester key. 

Changes here: https://github.com/matter-labs/era-contracts/pull/555/commits/9f9fb5bed614c4c105eff8d79120232ef76cf043